persistence/sql/sql_process_store.go

```diff
@@ -91,28 +92,372 @@ func (p sqlProcessStoreImpl) StopProcess(ctx context.Context, request persistenc
+	latestProcessExecutions, err := tx.SelectLatestProcessExecutionForUpdate(ctx, request.Request.Namespace, request.Request.ProcessId)
+	if err != nil {
+		return nil, err
+	}
+	if len(latestProcessExecutions) > 1 {
+		return nil, fmt.Errorf("more than one latest process execution for process id %v", request.Request.ProcessId)
+	}
+
+	// if it is still running, return already started
+	// if finished, start a new process
+	// if there is no previous run with the process id, start a new process
+	if len(latestProcessExecutions) == 1 {
+		processExecutionRowForUpdate, err := tx.SelectProcessExecutionForUpdate(ctx, latestProcessExecutions[0].ProcessExecutionId)
+		if err != nil {
+			p.logger.Error(err.Error())
+			return nil, err
+		}
+		if processExecutionRowForUpdate.Status == persistence.ProcessExecutionStatusRunning {
+			return &persistence.StartProcessResponse{
+				AlreadyStarted: true,
+			}, nil
+		}
+
+		prcExeId := uuid.MustNewUUID()
+		err = tx.UpdateLatestProcessExecution(ctx, extensions.LatestProcessExecutionRow{
+			Namespace:          request.Request.Namespace,
+			ProcessId:          request.Request.ProcessId,
+			ProcessExecutionId: prcExeId,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		hasNewWorkerTask, err := p.insertProcessExecution(ctx, tx, request, prcExeId)
+		if err != nil {
+			return nil, err
+		}
+
+		return &persistence.StartProcessResponse{
+			ProcessExecutionId: prcExeId,
+			AlreadyStarted:     false,
+			HasNewWorkerTask:   hasNewWorkerTask,
+		}, nil
+	}
+
+	prcExeId := uuid.MustNewUUID()
+	err = tx.InsertLatestProcessExecution(ctx, extensions.LatestProcessExecutionRow{
+		Namespace:          request.Request.Namespace,
+		ProcessId:          request.Request.ProcessId,
 		ProcessExecutionId: prcExeId,
 	})
+	if err != nil {
+		return nil, err
+	}
+
+	hasNewWorkerTask, err := p.insertProcessExecution(ctx, tx, request, prcExeId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &persistence.StartProcessResponse{
+		ProcessExecutionId: prcExeId,
+		AlreadyStarted:     false,
+		HasNewWorkerTask:   hasNewWorkerTask,
+	}, nil
+}
 
+func (p sqlProcessStoreImpl) applyAllowIfPreviousExitAbnormallyPolicy(
+	ctx context.Context,
+	tx extensions.SQLTransaction,
+	request persistence.StartProcessRequest) (*persistence.StartProcessResponse, error) {
+	latestProcessExecutions, err := tx.SelectLatestProcessExecutionForUpdate(ctx, request.Request.Namespace, request.Request.ProcessId)
 	if err != nil {
-		if p.session.IsDupEntryError(err) {
-			// TODO support other ProcessIdReusePolicy on this error
+		return nil, err
+	}
+	if len(latestProcessExecutions) > 1 {
+		return nil, fmt.Errorf("more than one running process execution for process id %v", request.Request.ProcessId)
+	}
+
+	if len(latestProcessExecutions) == 1 {
+		processExecutionRowForUpdate, err := tx.SelectProcessExecutionForUpdate(ctx, latestProcessExecutions[0].ProcessExecutionId)
```

Comment:
We need to add a other `SelectProcessExecution` (no ForUpdate) for ALLOW_IF_NO_RUNNING, ALLOW_IF_PREIVOUS_EXIT_ABNOARMALLY because they are not updating the current execution. This will save the locking contention.